File: bindings/node/__tests__/timemachine.test.js
"use strict";

const { test } = require("node:test");
const assert = require("node:assert");
const { TimeMachine } = require("../index.js");

function feed() {
  const lines = [
    {
      ts: 10,
      symbol: "BTC-USDT",
      feed: {
        kind: "market",
        type: "trade",
        symbol: { base: "BTC", quote: "USDT" },
        price: "100",
        quantity: "1",
        aggressor: "Buy",
        timestamp: 10,
      },
    },
    {
      ts: 20,
      symbol: "BTC-USDT",
      feed: {
        kind: "market",
        type: "trade",
        symbol: { base: "BTC", quote: "USDT" },
        price: "110",
        quantity: "2",
        aggressor: "Sell",
        timestamp: 20,
      },
    },
    { ts: 20, symbol: "ETH-USDT", feed: { kind: "funding", rate: 0.0002, mark_price: 50.0 } },
  ];
  return lines.map((l) => JSON.stringify(l)).join("\n");
}

function loaded() {
  const tm = new TimeMachine("{}");
  tm.command(JSON.stringify({ cmd: "load", data: feed() }));
  return tm;
}

test("seek reconstructs the snapshot", () => {
  const snap = JSON.parse(loaded().command(JSON.stringify({ cmd: "seek", ts: 20 })));
  assert.strictEqual(snap.ts, 20);
  assert.ok(Math.abs(snap.symbols["BTC-USDT"].last - 110.0) < 1e-9);
  assert.ok(snap.symbols["ETH-USDT"].funding !== null);
});

test("the same seek yields byte-identical output", () => {
  const cmd = JSON.stringify({ cmd: "seek", ts: 20 });
  assert.strictEqual(loaded().command(cmd), loaded().command(cmd));
});

test("an invalid spec throws", () => {
  assert.throws(() => new TimeMachine("{ not valid json"));
});

test("version is a string", () => {
  assert.strictEqual(typeof new TimeMachine("{}").version(), "string");
});

module.exports = { feed };

File: bindings/java/src/test/java/org/wickra/timemachine/TimeMachineTest.java
package org.wickra.timemachine;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class TimeMachineTest {
    // A deterministic two-record feed on SYM: a rising then falling trade.
    static final String FEED =
            "{\"ts\":10,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\","
                    + "\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"100\","
                    + "\"quantity\":\"1\",\"aggressor\":\"Buy\",\"timestamp\":10}}\n"
                    + "{\"ts\":20,\"symbol\":\"SYM\",\"feed\":{\"kind\":\"market\",\"type\":\"trade\","
                    + "\"symbol\":{\"base\":\"AAA\",\"quote\":\"USDT\"},\"price\":\"105\","
                    + "\"quantity\":\"1\",\"aggressor\":\"Sell\",\"timestamp\":20}}";

    // The feed embedded as a JSON string literal for the load command.
    static String loadCmd() {
        StringBuilder sb = new StringBuilder("{\"cmd\":\"load\",\"data\":\"");
        for (int i = 0; i < FEED.length(); i++) {
            char c = FEED.charAt(i);
            if (c == '"') {
                sb.append("\\\"");
            } else if (c == '\n') {
                sb.append("\\n");
            } else {
                sb.append(c);
            }
        }
        sb.append("\"}");
        return sb.toString();
    }

    static TimeMachine loaded() {
        TimeMachine tm = new TimeMachine("{}");
        tm.command(loadCmd());
        return tm;
    }

    @Test
    void versionIsNonEmpty() {
        assertFalse(TimeMachine.version().isEmpty());
    }

    @Test
    void seekReconstructsSnapshot() {
        try (TimeMachine tm = loaded()) {
            String out = tm.command("{\"cmd\":\"seek\",\"ts\":20}");
            assertTrue(out.contains("\"ts\":20"), out);
            assertTrue(out.contains("\"last\":105"), out);
        }
    }

    @Test
    void invalidSpecThrows() {
        assertThrows(IllegalArgumentException.class, () -> new TimeMachine("{ not valid json"));
    }
}
